closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -261,13 +265,38 @@ static void printErrors(void)
     fflush(stderr);
 }
 
+/*
+ * We use a 8 digit map (AABBCCDD) to represent the version of openssl.
+ * AA is the major version,
+ * BB is the minor version,
+ * CC is the revision, and DD for the letter that could be present in any version.
+ * For example, if an openssl version is in this scheme 1.1.1k.
+ * The ossl_ver should be 10101011 ( k is the 11th letter value for DD here... ).
+ */
+static jlong extractVersionToJlong(const char *astring)
+{
+    long major = 0;
+    long minor = 0;
+    long revision = 0;
+    long diff = 0;
+    char last_char = 0;
+    if (sscanf(astring, "OpenSSL %ld.%ld.%ld%c", &major, &minor, &revision, &last_char) < 3) {
+        return -1;
+    }
+    last_char = tolower(astring[strlen(astring) - 1]);
+    if (isalpha(last_char)) {
+        diff = (last_char - 'a') + 1;
+    }
```

Comment:
This still gets the wrong character. We want (the lowercase version of) the one seen by `sscanf()`.
Also, letters are not contiguous in EBCDIC, so I think we want the following which works in both ASCII and EBCDIC:
```c
    last_char = tolower(last_char);
    if (('a' <= last_char) && (last_char <= 'i')) {
        diff = last_char - 'a' + 1;
    } else if (('j' <= last_char) && (last_char <= 'r')) {
        diff = last_char - 'j' + 10;
    } else if (('s' <= last_char) && (last_char <= 'z')) {
        diff = last_char - 's' + 19;
    }
```

Reply:
Edited.